code/modules/garbage collection/garbage_collector.dm

```diff
@@ -0,0 +1,166 @@
+// Main garbage collection code
+
+// For general information about how the GC works and how to use it, see __gc_info.dm
+
+#define GC_COLLECTIONS_PER_TICK 300 // Was 100.
+#define GC_COLLECTION_TIMEOUT (30 SECONDS)
+#define GC_FORCE_DEL_PER_TICK 60
+//#define GC_DEBUG
+
+// A list of types that were queued in the GC, and had to be soft deleted; used in testing
+var/list/gc_hard_del_types = list()
+var/datum/garbage_collector/garbageCollector
+
+// The time a datum was destroyed by the GC, or null if it hasn't been
+/datum/var/gcDestroyed
+// Whether a datum was hard-deleted by the GC; 0 if not, 1 if it was queued, -1 if directly deleted
+/datum/var/hard_deleted = 0
+
+/datum/garbage_collector
+	var/list/queue = new
+	var/del_everything = 0
+
+	// To let them know how hardworking am I :^).
+	var/dels_count = 0
+	var/hard_dels = 0
+	var/soft_dels = 0
+
+/datum/garbage_collector/proc/addTrash(var/datum/D)
+	if(!istype(D) || del_everything)
+		del(D)
+		hard_dels++
+		dels_count++
+		return
+
+	queue -= "\ref[D]" // If this is a re-used ref, remove the old ref from the queue
+	queue["\ref[D]"] = world.time
+
+/datum/garbage_collector/proc/process()
+	var/remainingCollectionPerTick = GC_COLLECTIONS_PER_TICK
+	var/remainingForceDelPerTick = GC_FORCE_DEL_PER_TICK
+	var/collectionTimeScope = world.time - GC_COLLECTION_TIMEOUT
+	while(queue.len && --remainingCollectionPerTick >= 0)
+		var/refID = queue[1]
+		var/destroyedAtTime = queue[refID]
+
+		if(destroyedAtTime > collectionTimeScope)
+			break
+
+		var/datum/D = locate(refID)
+		// If the object still exists, and it's the same object, hard del it
+		if(D && D.gcDestroyed == destroyedAtTime)
+			if(remainingForceDelPerTick <= 0)
+				break
+
+			#ifdef GC_DEBUG
+			gcwarning("GC process force delete [D.type]")
```

Comment:
Preprocessor directives, like #ifdef, are handled before the code is actually compiled, so they have no effect on the indentation on their own.